src/internal/obj/util.go

```diff
@@ -28,24 +29,33 @@ func NewTestClient(t testing.TB) (Client, string) {
 
 // Copy copys an object from src at srcPath to dst at dstPath
 func Copy(ctx context.Context, src, dst Client, srcPath, dstPath string) (retErr error) {
-	rc, err := src.Reader(ctx, srcPath, 0, 0)
-	if err != nil {
-		return err
-	}
-	defer func() {
-		if err := rc.Close(); retErr == nil {
-			retErr = err
+	pr, pw := io.Pipe()
+	eg := errgroup.Group{}
+	eg.Go(func() (retErr error) {
+		if err := src.Get(ctx, srcPath, 0, 0, pw); err != nil {
+			pw.CloseWithError(err)
+			return err
 		}
-	}()
-	wc, err := dst.Writer(ctx, dstPath)
-	if err != nil {
-		return err
-	}
-	defer func() {
-		if err := wc.Close(); retErr == nil {
-			retErr = err
+		return pw.Close()
+	})
+	eg.Go(func() error {
+		return dst.Put(ctx, dstPath, pr)
+	})
+	return eg.Wait()
+}
+
+// With Pipe calls rcb with a reader and wcb with a writer
+func WithPipe(wcb func(w io.Writer) error, rcb func(r io.Reader) error) error {
+	pr, pw := io.Pipe()
+	eg := errgroup.Group{}
+	eg.Go(func() error {
+		if err := wcb(pw); err != nil {
+			return pw.CloseWithError(err)
 		}
-	}()
-	_, err = io.Copy(wc, rc)
-	return err
+		return pw.Close()
+	})
+	eg.Go(func() error {
+		return rcb(pr)
```

Comment:
I think we need a close with error for the read side as well. That should be enough to make it such that we don't need to thread a context through this helper function.